stablehlo/dialect/StablehloOps.cpp

```diff
@@ -1025,6 +1025,63 @@ LogicalResult DotOp::verify() {
   return success();
 }
 
+// Certain enum values have custom printers that include `<` `>`
+// i.e. `<PHILOX>` This method parses the enum value without `<` `>`.
+template <typename EnumAttrKind, typename SymbolizeFn>
+ParseResult parseRawEnum(OpAsmParser& parser, Attribute& enumAttr,
+                         llvm::StringRef errStr, SymbolizeFn symbolizeFn) {
+  llvm::SMLoc loc = parser.getCurrentLocation();
+  llvm::StringRef enumStr;
+  if (failed(parser.parseOptionalKeyword(&enumStr)))
+    return parser.emitError(loc) << "expected " << errStr << " enum";
+  auto enumValOpt = symbolizeFn(enumStr);
+  if (!enumValOpt.has_value())
+    return parser.emitError(loc)
+           << "invalid " << errStr << " enum value " << enumStr;
+  enumAttr = EnumAttrKind::get(parser.getContext(), enumValOpt.value());
+  return success();
+}
+
+// PrecisionConfig - Optional attribute, print the array as raw enums
+//
+// {precision_config = [#stablehlo<precision DEFAULT>,
+//                      #stablehlo<precision DEFAULT>]}
+// ==> ..., precision = [DEFAULT, DEFAULT]
+void printPrecisionConfig(OpAsmPrinter& p, Operation*,
+                          llvm::Optional<mlir::ArrayAttr> attrOpt) {
+  if (!attrOpt.has_value() || !attrOpt.value()) return;
```

Comment:
I don't think we should check for `!attrOpt.value()`. I would be surprised if an empty attribute (not an empty option, but an empty attribute) will be able to pass verification and end up here.

Reply:
I added that out of necessity (crashed without). Dug into the generated code a bit to see what was going on:

```
::mlir::ArrayAttr DotOp::getPrecisionConfigAttr() {
  return ::mlir::impl::getAttrFromSortedRange((*this)->getAttrs().begin() + 0, (*this)->getAttrs().end() - 0, getPrecisionConfigAttrName()).dyn_cast_or_null<::mlir::ArrayAttr>();
}

::llvm::Optional< ::mlir::ArrayAttr > DotOp::getPrecisionConfig() {
  auto attr = getPrecisionConfigAttr();
  return attr ? ::llvm::Optional< ::mlir::ArrayAttr >(attr) : (::llvm::None);
}

void DotOp::print(::mlir::OpAsmPrinter &_odsPrinter) {
  _odsPrinter << ' ';
  _odsPrinter << getLhs();
  _odsPrinter << ",";
  _odsPrinter << ' ';
  _odsPrinter << getRhs();
  printPrecisionConfig(_odsPrinter, *this, getPrecisionConfigAttr());
  _odsPrinter.printOptionalAttrDict((*this)->getAttrs(), /*elidedAttrs=*/{"precision_config"});
  _odsPrinter << ' ' << ":";
  _odsPrinter << ' ';
  _odsPrinter.printFunctionalType(getOperation()->getOperandTypes(), getOperation()->getResultTypes());
}
```

This is expected [according to documentation](https://mlir.llvm.org/docs/OpDefinitions/#custom-directives). I had misread the `Attribute Variables > Optional` section. The `ArrayAttr` gets passed to the print function as null. I was accidentally creating an optional of null with this function signature.

Fixed to use:

```
void printPrecisionConfig(OpAsmPrinter& p, Operation*, mlir::ArrayAttr attr) {
   if (!attr) return;
  // ....
}
```